plonky2-verifier/src/fields/goldilocks_quadratic.cairo

```diff
@@ -0,0 +1,152 @@
+use core::option::OptionTrait;
+use core::result::ResultTrait;
+use core::traits::TryInto;
+use core::integer::U128MulGuarantee;
+
+use plonky2_verifier::fields::goldilocks::{Goldilocks, gl, GoldilocksImpl};
+
+#[derive(Copy, Drop, Debug, PartialEq, Eq)]
+pub struct GoldilocksQuadratic {
+    pub a: Goldilocks,
+    pub b: Goldilocks,
+}
+
+#[generate_trait]
+pub impl GoldilocksQuadraticImpl of GoldilocksQuadraticTrait {
+    fn reduce_u64(val: u64) -> GoldilocksQuadratic {
+        GoldilocksQuadratic { a: GoldilocksImpl::reduce_u64(val), b: Goldilocks { inner: 0 }, }
+    }
+
+    fn reduce_u256(val: u256) -> GoldilocksQuadratic {
+        GoldilocksQuadratic { a: GoldilocksImpl::reduce_u256(val), b: Goldilocks { inner: 0 }, }
+    }
+
+    fn reduce_u128(val: u128) -> GoldilocksQuadratic {
+        GoldilocksQuadratic { a: GoldilocksImpl::reduce_u128(val), b: Goldilocks { inner: 0 }, }
+    }
+
+    fn add_u64(ref self: GoldilocksQuadratic, val: u64) -> GoldilocksQuadratic {
+        let other = GoldilocksQuadraticTrait::reduce_u64(val);
+        self + other
+    }
+}
+
+pub impl GoldilocksQuadraticAdd of core::traits::Add<GoldilocksQuadratic> {
+    fn add(lhs: GoldilocksQuadratic, rhs: GoldilocksQuadratic) -> GoldilocksQuadratic {
+        GoldilocksQuadratic { a: lhs.a + rhs.a, b: lhs.b + rhs.b }
+    }
+}
+pub impl GoldilocksQuadraticSub of core::traits::Sub<GoldilocksQuadratic> {
+    fn sub(lhs: GoldilocksQuadratic, rhs: GoldilocksQuadratic) -> GoldilocksQuadratic {
+        lhs + (-rhs)
+    }
+}
+pub impl GoldilocksQuadraticMul of core::traits::Mul<GoldilocksQuadratic> {
+    fn mul(lhs: GoldilocksQuadratic, rhs: GoldilocksQuadratic) -> GoldilocksQuadratic {
+        let W = gl(7); // Element W of BaseField, such that `X^d - W` is irreducible over BaseField.
+
+        let a0 = lhs.a;
+        let a1 = lhs.b;
+
+        let b0 = rhs.a;
+        let b1 = rhs.b;
+
+        let c0 = a0 * b0 + W * a1 * b1;
+        let c1 = a0 * b1 + a1 * b0;
+
+        GoldilocksQuadratic { a: c0, b: c1 }
+    }
+}
+
+pub impl GoldilocksQuadraticZero of core::num::traits::Zero<GoldilocksQuadratic> {
+    fn zero() -> GoldilocksQuadratic {
+        GoldilocksQuadratic { a: gl(0), b: gl(0) }
+    }
```

Comment:
Minor nitpicking: lets use the static constructor for this instead of the reduce call (ie. goldilocks {inner: 0})